### PR TITLE
ci(actions): action to add comment when blocking issue is closed

### PR DESCRIPTION
**Pull request by DitwanP**
**Related Issue:** #9722 

## Summary
Adds new action to create a comment on blocked issues when the blocking issue has been closed out. 

With these changes we would manually add blocked issues to the body of the blocking issue in the format:
 `Blocked issues: #1234, #5678...`
 
When closed the action would get those issues and add a comment saying the blocking issue has been closed as well removing blocked label (maybe).

### CODE REVIEW

File: .github/scripts/addUnblockedComment.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// When a blocking issue is closed:
+// 1. Leaves a comment on all the issues listed as blocked in body,
+
+module.exports = async ({ github, context }) => {
```
Comment (benelan) [inline .github/scripts/addUnblockedComment.js:4]:
```suggestion
/** @param {import('github-script').AsyncFunctionArguments} AsyncFunctionArguments */
module.exports = async ({ github, context }) => {
  const { repo, owner } = context.repo;

  const payload = /** @type {import('@octokit/webhooks-types').IssuesEvent} */ (context.payload);
module.exports = async ({ github, context }) => {
```

Can you please add types via JSDOC comments? See the other scripts in that directory for full examples

File: .github/scripts/addUnblockedComment.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  if (!issueBody) {
+    console.log("No issue body was found");
+    return;
+  }
+
+  // Get the list of issues blocked by this issue if there are any
+  const blockedIssuesLine = issueBody.match(blockedIssuesRegex);
+
+  // Add a "@" character to notify the user
+  const verifiers = ISSUE_VERIFIERS.split(",").map((v) => " @" + v.trim());
```
Comment (benelan) [inline .github/scripts/addUnblockedComment.js:18]:
```suggestion
  const verifiers = ISSUE_VERIFIERS.split(",").map((v) => `@${v.trim()}`);
```
Nitpick: can you use a string literal here instead of concatenating

File: .github/scripts/addUnblockedComment.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+  // If "Blocked issues" line is matched in the body then create a comment on each issue listed
+  if (blockedIssuesLine) {
+    const issueNumbers = blockedIssuesLine[0]
+      .slice(16)
+      .split(", ")
+      .map((number) => number.slice(1));
```
Comment (benelan) [inline .github/scripts/addUnblockedComment.js:25]: The regex isn't strict enough for this match parsing. For example, the regex would match this string, which would break the slice logic:

> Blocked issues:           #123,#321

File: .github/scripts/addUnblockedComment.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    for (const issueNumber of issueNumbers) {
+      const issueProps = {
+        issue_number: issueNumber,
+        owner: context.repo.owner,
+        repo: context.repo.repo,
+      };
+
+      await github.rest.issues.createComment({
+        ...issueProps,
+        body: `Issue #${context.issue.number} has been closed, this issue can now move forward. cc ${verifiers}`,
```
Comment (benelan) [inline .github/scripts/addUnblockedComment.js:36]: Would it be possible for an issue to have multiple blockers? If so, this message might give the issue author false hope if other blockers are still open. Maybe we could rephrase it to something like:

> The blocking issue #${context.issue.number} has been closed. This issue is ready for re-evaluation, cc ${verifiers}.

File: .github/scripts/addUnblockedComment.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        if (error.status === 404) {
+          console.log(`The label "blocked" does not exist on issue #${issueNumber}.`);
+        } else {
+          throw error;
+        }
+      }
+    }
+  } else {
+    console.log(`No blocked issues listed in the body of issue #${context.payload.issue.number}.`);
+    return;
```
Comment (benelan) [inline .github/scripts/addUnblockedComment.js:54]: Nitpick: this `return` is unnecessary since it's at the end of the script

**Review by benelan — CHANGES_REQUESTED** [review]
I added a couple comments but it's almost there, nice work Ditwan!

File: .github/scripts/addUnblockedComment.js (hunk 4 above)
Comment (DitwanP) [reply .github/scripts/addUnblockedComment.js:36]: Agreed!

File: .github/scripts/addUnblockedComment.js (hunk 3 above)
Comment (DitwanP) [reply .github/scripts/addUnblockedComment.js]: Nice! tweaked a bit to `/(?!Blocked issues:\s)(#\d+)/gi` so it doesn't include the commas at the end of the numbers ["#123","#321","#11"]

Comment (benelan) [reply .github/scripts/addUnblockedComment.js]: Nice catch, it was late and they must have blended in with the array commas lol

File: .github/scripts/addUnblockedComment.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+  // If "Blocked issues" line is matched in the body then create a comment on each issue listed
+  if (blockedIssues) {
+    const issueNumbers = blockedIssues.map((number) => number.slice(1));
+
+    for (const issueNumber of issueNumbers) {
+      const issueProps = {
+        repo,
+        owner,
+        issue_number: issueNumber,
```
Comment (benelan) [inline .github/scripts/addUnblockedComment.js:32]: I'm seeing type errors for the `createComment` and `removeLabel` function calls. Are you not seeing this on your end?

> tsserver: Argument of type '{ body: string; repo: string; owner: string; issue_number: string; }' is not assignable to parameter of type 'RequestParameters & { owner: string; repo: string; issue_number: number; body: string; }'.
  Type '{ body: string; repo: string; owner: string; issue_number: string; }' is not assignable to type '{ owner: string; repo: string; issue_number: number; body: string; }'.
    Types of property 'issue_number' are incompatible.
      Type 'string' is not assignable to type 'number'. [2345]
 
 I wonder if I need to fix something with the types setup. Regardless, `issueNumber` needs to be a `number`, not a `string`.

```suggestion
        issue_number: Number(issueNumber),
```

**Review by benelan — APPROVED** [review]
Adding types uncovered one more problem, but after that's fixed it LGTM!
